specification commit v0.1 (#1)
hyperledger-identus/vdr: README.md

The system decouples the application layer from underlying storage mechanisms by leveraging pluggable components: **Drivers** and **URL Managers**. 
This specification defines the roles, relationships, and data model used by these components, as well as standard URL query parameters that facilitate driver selection and data integrity verification.

---

## 2. Introduction

The VDR system is designed to enable verifiable data operations across various storage backends such as databases, blockchains, or in-memory storage. 
It also supports flexible URL construction and parsing for addressing stored data. 
Through its modular architecture, the system can adapt to different technologies and deployment environments.

Key components include:

- **VDR Proxy**: The entry point that mediates client requests and delegates operations.
- **Drivers**: Plugins that implement the actual data storage and retrieval logic.
- **URL Managers**: Modules that construct and resolve URLs carrying essential metadata.

Future enhancements include a centralized registry for Drivers and URL Managers to facilitate dynamic discovery and management.

---

> ## 3. Actors

Reviewer: I believe you are missing one actor. The VDR provider it self.
The issuers can be the one you has the instance of the VDR. But can also use the VDR provided by another actor.

Author: You mean an actor that just provides access to the VDR?

Reviewer: yes

Author: Dont really agree much that's an "actor".

Author: Updated

  The diversity of storage technologies—from centralized databases to distributed ledger systems and in-memory caches—creates fragmentation. A solution is needed that can abstract these differences and provide a unified, interoperable interface for data storage and retrieval. Such a solution must be flexible enough to accommodate various storage backends while ensuring that the verification process remains consistent across implementations.

- **Scalability and Trustless Collaboration:**  
  In many modern applications, data is produced by one party and consumed by many. Ensuring that each consumer can independently verify the data without relying on a centralized intermediary is essential. A scalable solution must support trustless collaboration, where the integrity and provenance of data can be verified by all stakeholders without compromising performance or security.

### 4.2 How VDR Addresses These Challenges

The Verifiable Data Registry (VDR) is designed to solve these complex problems by providing a modular, extensible framework that decouples data operations from the underlying storage mechanisms. Its architecture supports the following:

> - **Public Availability and Decentralization:**  

Reviewer: But this depends on the VDR driver.
You can't clam this about the VDR and talk about in-memory storage drivers...

Author: Fixed

- **Extensible Metadata and Query Registries:**  
  The system introduces registries for both URL query parameters and metadata keys. These registries ensure that all necessary information for verifying and retrieving data (such as driver identifiers, families, and versions) is consistently used across the system. This extensibility makes it easier to introduce new verification mechanisms and adapt to emerging data integrity challenges.

- **Trustless Environment for Multi-party Verification:**  
  The VDR supports an ecosystem where multiple parties can independently verify the authenticity of stored data. This trustless environment reduces the reliance on any single party for data verification, enhancing security and fostering collaboration in scenarios where data integrity is paramount.

---

## 5. Terminology

The following terms are used throughout this specification:

> - **VDR**: Verifiable Data Registry – a system that offers a unified API for data operations.

Reviewer: Maybe not the best definition

Author: What would you say the definition would be?

Reviewer: I would adapt something close to the definition of other people.
"A verifiable data registry is a role a system plays to manage data related to the creation and verification of verifiable credentials." and "include trusted databases, decentralized databases, government ID databases, and distributed ledgers."

Author: That's in the context of verifiable credentials. But I get your point let me check other definitions that are more general.

Author: Updated

Additionally, Drivers are responsible for handling errors internally and for communicating any issues by setting the appropriate state in the `OperationResult`. 
For mutable data, Drivers take on the additional responsibility of managing and resolving updated data records. 
They must determine how to locate and retrieve the latest version of data when a URL points to a record that has undergone mutation.
Furthermore, the Driver is also responsible for the proof mechanism within the system. 
When a verification request is made, the Driver performs the necessary cryptographic operations—such as computing a SHA256 hash for immutable data or verifying a digital signature for mutable data—to generate a `Proof` object. 
This proof contains a type identifier, the cryptographic proof itself, and, if requested, the data used in the verification process. 
By handling the proof mechanism, the Driver ensures that clients can independently verify the integrity and authenticity of the stored data, thus reinforcing trust in the system.

- Implements the actual storage mechanism (e.g., database, blockchain, in-memory).

> - Performs operations such as `create`, `update`, `read`, and `delete`.

Reviewer: What about the operation to get the status of a entry?

- Returns a `OperationResult` that includes operational metadata (paths, queries, fragments, etc.) and the operation state.
- Handles errors internally and communicates them through the returned operation state.
- **Mutable Data Handling**: It is the responsibility of the Driver to manage and resolve data that has been mutated. This includes determining how to locate and retrieve updated versions of the data when a URL points to data that has changed.
- **Driver Family**: Drivers within the same family share common standards for URL interpretation and data handling. This ensures that the same URL can be correctly processed by any Driver in the family, thereby retrieving consistent data across different implementations.

> **Driver Family:**

Author: Example:

We got 3 drivers of Cardano:

- Cardano Driver1(Dr1) (with local connection to node)
- Cardano Driver2(Dr2) (with API to connect with other Cardano Drivers)
- Cardano Driver3(Dr3) (with connection to other nodes but does does not sync)

Issuer creates data with Dr1, version 1.0 of the family `cardano` and shares the URL (`did:example:abc/cardano/Dr1/1.0?block=<hash>&blockSlot=<slot>&transaction=<hash>&h=<integrity hash>&m=0`)

The drivers Dr2 and Dr3 are of the same family and any of them can process the URL and return the data.

# VDR System Technical Specification

This document specifies the architecture, components, and data model for the Verifiable Data Registry (VDR) system. It provides technical details for implementers and integrators.

---

## Authors and Contributors

This document is developed and maintained by a dedicated team of authors and editors committed to advancing the Verifiable Data Registry (VDR) system. 
Contributions from the community are welcome and encouraged. 
Please refer to our [Contribution Guidelines](https://github.com/hyperledger-identus/vdr/CONTRIBUTING.md) for further details on how you can contribute.

### Authors

- **Gonçalo Frade** <br>
  IOG Engineer <br>
  [goncalo.frade@iohk.io](mailto:goncalo.frade@iohk.io) <br>
  [github](https://github.com/goncalo-frade-iohk)

> 

Reviewer: ```suggestion
- **Fabio Pinheiro** <br>
  IOG Engineer <br>
  [fabio.pinheiro@iohk.io](mailto:fabio.pinheiro@iohk.io) <br>
  [github](https://github.com/FabioPinheiro)
```

| **urltyp**       | URL Type          | Identifies the specific URL Type to use. If only one URL manager is available this key is not required.            |
| **m**            | Mutable           | A Boolean indicating the data can be modified.                                                                     |

### 7.3 OperationResult

Drivers return a `OperationResult` object after performing storage operations. This object includes:

- **identifier**: A unique ID for the operation.
- **state**: The operation state (e.g., SUCCESS, ERROR).
- **paths**: An array representing storage paths.
- **queries**: A map containing query parameters.
- **fragment**: An optional URL fragment.
- **publicKeys**: An optional array of public keys related to the stored data.

### 7.4 Proof

The VDR system provides a robust mechanism to verify the integrity and authenticity of stored data through a cryptographic proof. 
The `Proof` object encapsulates all the necessary information to validate the data without exposing the raw content itself, unless explicitly requested. 

> This mechanism is fundamental to establishing trust within the system by allowing any consumer to independently verify that the data remains unaltered from its original state.

Reviewer: ```suggestion
This mechanism is fundamental to establishing trust within the system by allowing any consumer to independently verify that the data remains untremped from third parties.
```

Are you talking just about inmutable data?

By incorporating multiple types of cryptographic proofs, the VDR system can meet diverse security requirements while maintaining efficiency. 
Immutable data can be verified quickly using simple hashes, whereas mutable data benefits from the stronger assurances provided by digital signatures and advanced proof techniques such as Merkle trees. 
Drivers play a crucial role in this process, as they are responsible for generating and managing these proofs. 
They must implement the appropriate cryptographic mechanisms based on the nature of the data and the specific requirements of the operation.

This flexibility is essential for supporting a wide range of applications—from financial records and digital identities to supply chain management and open government data—by allowing each Driver to choose and execute the most suitable proof mechanism for its storage medium.

Overall, the proof mechanism is a cornerstone of the VDR system, ensuring that every piece of stored data can be independently validated. 

> The responsibility of generating, maintaining, and providing these proofs lies with the Drivers, which in turn fosters a secure, transparent, and decentralized ecosystem where data integrity and authenticity are maintained through robust cryptographic verification.

Reviewer: The thing of saying that the responsibility of generating the proof he lies on the driver means that the driver is access to the private keys.
Which means this VDR has access to the private kids.

Reviewer: The issuer should run the VDR's Driver itself to generate those proofs IMO

Author: The VDR does not have access to the private keys, is the driver responsibility to define how it will manage private keys.

Reviewer: yeah I need the consideration you make after on point 10.3

Reviewer 2: I'm also concerned about sending the private keys to VDR or Driver.
I propose to have some kind of RequestBuilder (or whatever) that acceps the Signer of particular type, or, as alternative, make a protocol between the user <-> vdr <-> driver that allows used to sign the data.

Author: It is well defined that this is on the Driver scope not the VDR, as we talked before we removed from the VDR API any thing with private keys

- **Cloud-Agent Enhancement:**  
  Extend the cloud-agent integration to support the Cardano Blockchain Driver. This will enable multi-party verification and trustless data sharing by leveraging the inherent security and decentralization of the blockchain.

These milestones represent significant steps toward a fully functional and extensible VDR system that supports a variety of storage backends and integrates seamlessly with decentralized agent architectures.
---

## 10. Security Considerations

The VDR system is designed to foster a trustless and decentralized ecosystem, but several security aspects must be carefully considered to ensure the system's integrity and reliability:

1. **Data Integrity and Authenticity**  
   The system relies on cryptographic proofs—such as hashes for immutable data and digital signatures for mutable data—to verify that data has not been tampered with. It is critical to ensure that these cryptographic algorithms are implemented correctly and that the keys used for signing remain secure. Any compromise of these keys could allow an attacker to forge proofs and undermine trust in the data.

2. **Driver and URL Manager Trustworthiness**  

>    Since Drivers and URL Managers are responsible for processing and constructing the metadata and proofs, it is essential to guarantee that these components are secure and properly validated. Malicious or poorly implemented Drivers may return incorrect results or expose sensitive information. Implementations should adhere to secure coding practices and undergo rigorous security audits.

Reviewer: The driver should be able to run on a secure environment or any environment, controlled by the issue when generating proofs.

Reviewer: This is +- point 3

Author: Is there any action on this or is it resolved? :D

- **Driver**: A plugin responsible for implementing data storage, mutation, retrieval, and removal.
- **URL Manager**: A component that constructs and resolves URLs used to reference stored data.
- **OperationResult**: The outcome of a storage operation, containing metadata such as operation state, paths, queries, and fragments.
- **Query Parameters Registry**: A registry of URL components that carry metadata for driver selection and data integrity verification.
- **Metadata Registry**: A registry of metadata keys used throughout the system. This registry is designed to be extensible, similar to the URL query parameters registry.

---

# 6. Architectural Overview

The VDR system is composed of several core components that work together to perform data operations. In this section, we detail how the components interact for each primary task: storing, mutating, retrieving, and removing data, with special consideration for handling mutable data.

## 6.1 Component Responsibilities

> ### 6.1.1 VDR Proxy (VDR Interface)

Reviewer 2: ```suggestion
### 6.1.1 VDR Interface
```

Author: Removed all VDR Proxy and replaced by VDR Interface

The Driver Interface is at the heart of the actual data operations within the VDR system.
Drivers are the pluggable components that implement the storage mechanisms, which could range from traditional databases to blockchain-based systems or in-memory storage solutions.
When a client request is delegated by the VDR Proxy, the appropriate Driver is tasked with executing the requested operation.
This includes creating new data entries, updating existing data, retrieving stored data, or deleting data.
Each Driver returns an `OperationResult` object that encapsulates detailed information about the operation, including storage paths, query parameters, fragments, and any relevant public keys.
Additionally, Drivers are responsible for handling errors internally and for communicating any issues by setting the appropriate state in the `OperationResult`.
For mutable data, Drivers take on the additional responsibility of managing and resolving updated data records.
They must determine how to locate and retrieve the latest version of data when a URL points to a record that has undergone mutation.
Furthermore, the Driver is also responsible for the proof mechanism within the system.

> When a verification request is made, the Driver performs the necessary cryptographic operations—such as computing a SHA256 hash for immutable data or verifying a digital signature for mutable data—to generate a `Proof` object.

Reviewer 2: Given that the Holder/Verifier (the data consumer) calls the verification request, what is the `Proof` object format, and how can the Driver generate it?

Author: Its in the documentation what is the proof object in data models

Reviewer 2: The `Proof` object is usually associated with the JSON-LD spec and is a [W3C standard](https://w3c.github.io/vc-data-integrity/#proofs).
Are we defining a new spec for `Proof` object or going to follow the existing standard?

This proof contains a type identifier, the cryptographic proof itself, and, if requested, the data used in the verification process.

> By handling the proof mechanism, the Driver ensures that clients can independently verify the integrity and authenticity of the stored data, thus reinforcing trust in the system.

Reviewer 2: Shouldn't the Proof object or any other signature be stored with the initial data?

### 6.2.4 Importance of Proofs in a Mutable Environment

In environments where data mutability is required, the use of digital signatures and advanced cryptographic proofs is critical. They provide:

- **Security:** Ensuring that any data modification is authenticated and authorized by the original Issuer.
- **Transparency:** Allowing multiple parties to independently verify that the data has not been tampered with, even as it evolves.
- **Interoperability:** Enabling the system to integrate with various storage backends and proof standards while maintaining consistent verification methods.

Overall, the dual approach to proof generation—simple hashes for immutable data and digital signatures (potentially enhanced with Merkle trees or anoncreds proofs) for mutable data—ensures that the VDR system can support a wide array of applications. From financial records and digital identities to supply chain management and government data, every piece of stored data can be independently verified, thus upholding the integrity and trustworthiness of the entire ecosystem.

## 6.3 Data Flow by Task

### 6.3.1 Creating Data

> 1. **Client Request**: The client sends data and associated metadata to the VDR Interface. If a private key is included, it signifies that the data **may** be mutable.

Reviewer 3: If a private key is included data is mutable ?

Author: Good catch this is wrong. Ive fixed it ;)

2. **Driver Selection**:
- The VDR checks if a single or multiple Drivers are available.

> - If multiple Drivers exist, the VDR selects the appropriate one based on provided metadata (e.g., driver identifier, family, version).

Reviewer 3: How do i know which Driver has  priority over what should be the fallback mechanism

Author: When there is more than one driver plugin you need to pass the `drid` as metadata, its on the Metadata registry section ;)

5. **Response:**  
   The Driver returns a `Proof` object that contains the type of proof, optional data (if requested by the `returnData` flag), and the cryptographic proof itself. The VDR Interface then forwards this `Proof` to the client as confirmation of the data's integrity and authenticity.

```mermaid
sequenceDiagram
    participant C as Client
    participant V as VDR Interface
    participant U as URL Manager
    participant D as Driver

    C->>V: Request to verify data (providing URL, returnData flag)
    V->>U: Call resolve(URL)
    U-->>V: Return URL components (paths, queries, fragment, public keys)
    V->>D: Call verify(url, returnData) with resolved components
    D-->>V: Return Proof (type, optional data, cryptographic proof)
    V-->>C: Return verification proof (confirmation of data integrity/authenticity)
```

---

## 7. Data Model

### 7.1 URL Path
The VDR system enforces a standardized URL path structure to ensure consistency and conformance across all generated URLs. Each URL path must include the following segments in order:

```

> <base-url>/<driver family>/<driver identifier>/<driver version>/<custom paths>

Reviewer 3: Do we need  not any validation on custom paths ?

Author: What do you mean?

Reviewer 3: I mean custom paths  is not validated  will this not be  way to exploit <base-url>/driver-family/1234/1.0/../../../etc/passwd

Author: Yes after talking with Yurii we decided to go back having this in queries not as part of the path, Ive already committed the changes ;)

This mechanism is fundamental to establishing trust within the system by allowing any consumer to independently verify that the data remains untampered by third parties.

The `Proof` object includes the following properties:

- **type**:  
  A string that indicates the type of cryptographic proof generated. Examples include:
    - **SHA256**: A simple hash of the data, used when the data is immutable.
    - **EDDSAWithSHA256**: A digital signature combining the EDDSA algorithm with SHA256 hashing, typically used when the data is mutable.
    - **RSAWithSHA256**: A digital signature generated using RSA encryption alongside SHA256 hashing.
    - **ECDSAWithSHA256**: A signature using the Elliptic Curve Digital Signature Algorithm (ECDSA) in combination with SHA256.  
      These proof types enable the system to adapt to different security requirements and operational scenarios.

- **data**:  
  An optional byte array containing the retrieved data if the verification request specifies that the data should be returned (using the `returnData` flag). This allows verifiers to compare the actual data against the proof if necessary.

- **proof**:  

>   A byte array representing the cryptographic proof itself. Depending on the nature of the data, this may be a hash (for immutable data) or a digital signature (for mutable data). The proof is used by verifiers to ensure that the data has not been altered since it was originally stored.

Reviewer 3: What is the reason we expose byteArray without any encoding.?

Author: Uhn I see your point. But dont know how giving an encoding like bas64URL would help in this situation. Also the proof type can represent an encoding.

Author: IMO it can be raw binary or we can default to base64url?
